Source/Parallelization/WarpXComm.cpp

```diff
@@ -871,33 +873,143 @@ WarpX::SyncCurrent (
         }
     }
 
-    // Restrict fine patch current onto the coarse patch, before
-    // summing the guard cells of the fine patch
-    for (int lev = 1; lev <= finest_level; ++lev)
+    // If there is a single level, we apply the filter on the fp data and
+    // then call SumBoundary that adds data from different boxes. This needs
+    // to be done because a particle near a box boundary may deposit current
+    // at a given (i,j,k) that is on the edge or in the ghost region of that
+    // box while at the same time that (i,j,k) is also in the valid region
+    // of another box. After SumBoundary, the result is as if there is only
+    // a single box on a single process. Also note we need to call
+    // SumBoundary even if there is only a single process, because a process
+    // may have multiple boxes. Furthermore, even if there is only a single
+    // box on a single process, SumBoundary should also be called if there
+    // are periodic boundaries. So we always call SumBounary even if it
+    // might be a no-op in some cases, because the function does not perform
+    // any communication if not necessary.
+    //
+    // When there are multiple levels, we need to send data from fine levels
+    // to coarse levels. In the implementation below, we loop over levels
+    // from the finest to the coarsest. On each level, filtering and
+    // SumBoundary are done as the last two things. So the communication
+    // data on the sender side are always unfiltered and unsummed. The
+    // receivers are responsible for filtering that is dependent on the grid
+    // resolution. On the finest level, we coarsen the unsummed fp data onto
+    // cp grids, which are the coarsened version of the fp grids with the
+    // same DistributionMapping. Then on the level below, We use ParallelAdd
+    // to add the finer level's cp data onto the current level's fp
+    // data. After that, we apply filter and SumBounadry to the finer
+    // level's cp MultiFab. At this time, the finer level's fp and cp data
+    // have all been properly filtered and summed. For the current level, if
+    // there are levels below this, we need to process this level's cp data
+    // just like we have done for the finer level. The iteration continues
+    // until we reache level 0. There are however two additional
+    // complications.
+    //
+    // The first complication is that simply calling ParallelAdd to add the
+    // finer level's cp data to the current level's fp data does not
+    // work. Suppose there are multiple boxes on the current level (or just
+    // a single box with periodic bounaries). A given (i,j,k) can be present
+    // in more than one box. At the time of calling ParallelAdd, the current
```

Comment:
Only for nodal staggering of rho/j, right?
```suggestion
    // a single box with periodic bounaries). A given (i,j,k) can be present
    // in more than one box for nodal data in AMReX.
    // At the time of calling ParallelAdd, the current
```

Reply:
Right, only for non-cell-centered rho/j.